### PR TITLE
fix: [CDS-71719]: return valid capability if response code is less than 500

### PR DESCRIPTION
**Pull request by sarthakkasat**
## Describe your changes
For http step if the response code is less than 500 return it as valid capability

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)

### CODE REVIEW

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java [hunk 1]
```diff
@@ -39,14 +39,20 @@ public class HttpTaskParametersNg implements TaskParameters, ExecutionCapability
...
 
   @Override
   public List<ExecutionCapability> fetchRequiredExecutionCapabilities(ExpressionEvaluator maskingEvaluator) {
-    if (shouldAvoidHeadersInCapability) {
+    if (isIgnoreResponseCode && shouldAvoidHeadersInCapability) {
+      return Collections.singletonList(
+          HttpConnectionExecutionCapabilityGenerator.buildHttpConnectionExecutionCapabilityWithIgnoreResponseCode(
+              url, maskingEvaluator, true));
+    }
+    if (!isIgnoreResponseCode && shouldAvoidHeadersInCapability) {
```
Comment (sourabh-awashti) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java:55]: don't need to add isIgnoreResponseCode condition here

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java [hunk 2]
```diff
@@ -57,6 +63,11 @@ public List<ExecutionCapability> fetchRequiredExecutionCapabilities(ExpressionEv
         headers.add(KeyValuePair.builder().key(headerConfig.getKey()).value(headerConfig.getValue()).build());
       }
     }
+    if (isIgnoreResponseCode && !shouldAvoidHeadersInCapability) {
```
Comment (sourabh-awashti) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java:66]: don't need to add shouldAvoidHeadersInCapability condition here

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java (hunk 1 above)
Comment (sarthakkasat) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java]: Updated the code.

**Comment by yogesh-chauhan** [comment]
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java` line 55 at r3](https://reviewable.io/reviews/harness/harness-core/49204#-NXsMuPM0Nyiqk1bjqE2:-NXsMuPM0Nyiqk1bjqE3:bvsyoea) ([raw file](https://github.com/harness/harness-core/blob/aa1ab164bc5209cf56e64073e28d150a07e9c081/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java#L55)):*
> ```Java
>               url, maskingEvaluator, true));
>     }
>     if (shouldAvoidHeadersInCapability) {
> ```

```
if(shouldAvoid) {
if(ignoreResponseCode) {
       return Collections.singletonList(
          HttpConnectionExecutionCapabilityGenerator.buildHttpConnectionExecutionCapabilityWithIgnoreResponseCode(
              url, maskingEvaluator, true));

} else {
           return Collections.singletonList(
          HttpConnectionExecutionCapabilityGenerator.buildHttpConnectionExecutionCapability(
              url, QUERY, maskingEvaluator))
}
} else {
 // 
}
```

**Comment by yogesh-chauhan** [comment]
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/mixin/HttpConnectionExecutionCapabilityGenerator.java` line 50 at r3](https://reviewable.io/reviews/harness/harness-core/49204#-NXsOW5g1zXInVN7icSf:-NXsOW5g1zXInVN7icSg:b-ylt4ir) ([raw file](https://github.com/harness/harness-core/blob/aa1ab164bc5209cf56e64073e28d150a07e9c081/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/mixin/HttpConnectionExecutionCapabilityGenerator.java#L50)):*
> ```Java
> 
>   public static HttpConnectionExecutionCapability
>   buildHttpConnectionExecutionCapabilityWithIgnoreResponseCodeWithHeaders(
> ```

we don't need to have so much elaborate names. Its better to just use method overloading instead with a simple name
buildHttpExecCapability(..)

**Comment by yogesh-chauhan** [comment]
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 142 at r3](https://reviewable.io/reviews/harness/harness-core/49204#gh-1229003646:-NXsOd4G8tlVEdmFB01C:biwxjiw) ([raw file](https://github.com/harness/harness-core/blob/aa1ab164bc5209cf56e64073e28d150a07e9c081/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L142)):*
<details><summary><i>Previously, srinivaswings (Srinivasa Gurubelli) wrote…</i></summary><blockquote>

Can these both Feature flags be properties of HttpStep or Preferences/
</blockquote></details>

We are going to GA these soon

**Comment by sarthakkasat** [comment]
trigger uts

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java [hunk 3]
```diff
@@ -47,17 +48,31 @@ public class HttpTaskParametersNg implements TaskParameters, ExecutionCapability
...
-              url, QUERY, maskingEvaluator));
-    }
-    List<KeyValuePair> headers = new ArrayList<>();
-    if (EmptyPredicate.isNotEmpty(requestHeader)) {
-      for (HttpHeaderConfig headerConfig : requestHeader) {
-        headers.add(KeyValuePair.builder().key(headerConfig.getKey()).value(headerConfig.getValue()).build());
+      if (isIgnoreResponseCode) {
+        return Collections.singletonList(
+            HttpConnectionExecutionCapabilityGenerator.buildHttpConnectionExecutionCapabilityWithIgnoreResponseCode(
+                url, maskingEvaluator, true));
```
Comment (yogesh-chauhan) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java:54]: we should use the same HTTP details = QUERY here as the else block

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java (hunk 3 above)
Comment (sarthakkasat) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java:54]: Will Update

**Comment by yogesh-chauhan** [comment]
Please write unit tests for this change

**Review by hingerabhinav — COMMENTED** [review]
> Reviewed 2 of 3 files at r2, 2 of 2 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)* status: 5 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
*[`950-delegate-tasks-beans/src/test/java/io/harness/delegate/task/http/HttpTaskParametersNgTest.java` line 69 at r7](https://reviewable.io/reviews/harness/harness-core/49204#-NXzMGYJCRYOdPZjCBys:-NXzMGYJCRYOdPZjCByt:b-idjq3y) ([raw file](https://github.com/harness/harness-core/blob/dd0ee2e587bc13a7177f8f15b5ba7d8c6dfa9859/950-delegate-tasks-beans/src/test/java/io/harness/delegate/task/http/HttpTaskParametersNgTest.java#L69)):*
> ```Java
>             .build();
> 
>     List<ExecutionCapability> executionCapabilities = httpTaskParametersNg.fetchRequiredExecutionCapabilities(null);
> ```

Also assert on the size of execution capabilities

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 3 files at r2, 1 of 2 files at r6, 1 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)* status: 2 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained.

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 2 files at r6, 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)* status: 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
trigger uts

**Review by vaibhav-si — COMMENTED** [review]
> Reviewed 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)* status: Some of the checks are too old. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained.

**Comment by sarthakkasat** [comment]
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

Please write unit tests for this change
</blockquote></details>

Done.

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)* status: Some of the checks are too old. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java` line 66 at r2](https://reviewable.io/reviews/harness/harness-core/49204#gh-1228985183:-NYH9rX5ECT274ewJNaA:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/8047004eaa62f13126503e7618340c58abd752fa/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java#L66)):*
<details><summary><i>Previously, sourabh-awashti wrote…</i></summary><blockquote>

don't need to add shouldAvoidHeadersInCapability condition here
</blockquote></details>

Done.
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 142 at r3](https://reviewable.io/reviews/harness/harness-core/49204#gh-1229003646:-NYH9uCf88gp60aqL-S8:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aa1ab164bc5209cf56e64073e28d150a07e9c081/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L142)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

We are going to GA these soon
</blockquote></details>

Done.

**Review by hingerabhinav — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)* status: Some of the checks are too old. 1 change requests. 4. unresolved discussions. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)* status: Some of the checks are too old. 3. unresolved discussions.

**Comment by yogesh-chauhan** [comment]
trigger review

**Comment by sahilhindwani-harness** [comment]
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java` line 51 at r7](https://reviewable.io/reviews/harness/harness-core/49204#-NYHEKgw3PkKi74is9ZJ:-NYHEKgw3PkKi74is9ZK:bhjfgv8) ([raw file](https://github.com/harness/harness-core/blob/dd0ee2e587bc13a7177f8f15b5ba7d8c6dfa9859/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java#L51)):*
> ```Java
>   public List<ExecutionCapability> fetchRequiredExecutionCapabilities(ExpressionEvaluator maskingEvaluator) {
>     if (shouldAvoidHeadersInCapability) {
>       if (isIgnoreResponseCode) {
> ```

Please check if delegate release is required as there is a change in delegate contract.

**Comment by sahilhindwani-harness** [comment]
As discussed , please check the services that needs to be released if we are hotfixing this. Ideally this would require release of manager and pipeline-service but please ensure that things work as expected with older delegate (without this change) too.

**Comment by hingerabhinav** [comment]
@sahilhindwani-harness Pipeline service is calling the capability creation logic. So if this service has the updated params, we're picking the boolean up. Now that capability is created correctly, delegate has the code to handle that already.
We tested with older wings, ng manager and delegate and its working. 

Screenshot 2023-06-19 at 12 15 00 PM

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java [hunk 4]
```diff
@@ -47,17 +48,31 @@ public class HttpTaskParametersNg implements TaskParameters, ExecutionCapability
...
     if (shouldAvoidHeadersInCapability) {
-      return Collections.singletonList(
-          HttpConnectionExecutionCapabilityGenerator.buildHttpConnectionExecutionCapability(
-              url, QUERY, maskingEvaluator));
-    }
-    List<KeyValuePair> headers = new ArrayList<>();
-    if (EmptyPredicate.isNotEmpty(requestHeader)) {
-      for (HttpHeaderConfig headerConfig : requestHeader) {
-        headers.add(KeyValuePair.builder().key(headerConfig.getKey()).value(headerConfig.getValue()).build());
+      if (isIgnoreResponseCode) {
```
Comment (puthrayaharness) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java:51]: This whole code feels way too cluttered. There is too much nesting. Can we refactor the code for better readability? Consider having private methods.

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java [hunk 5]
```diff
@@ -47,17 +48,31 @@ public class HttpTaskParametersNg implements TaskParameters, ExecutionCapability
...
+      }
+    } else {
+      List<KeyValuePair> headers = new ArrayList<>();
+      if (EmptyPredicate.isNotEmpty(requestHeader)) {
+        for (HttpHeaderConfig headerConfig : requestHeader) {
+          headers.add(KeyValuePair.builder().key(headerConfig.getKey()).value(headerConfig.getValue()).build());
+        }
+      }
+      if (isIgnoreResponseCode) {
+        return Collections.singletonList(
```
Comment (puthrayaharness) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java:68]: How will headers impact this check?

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)* status: Some of the checks are too old. 1 change requests. 5. unresolved discussions. 

___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java` line 51 at r7](https://reviewable.io/reviews/harness/harness-core/49204#-NYHEKgw3PkKi74is9ZJ:-NYHXR8n0B4t1eBex28_:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/dd0ee2e587bc13a7177f8f15b5ba7d8c6dfa9859/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java#L51)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

Please check if delegate release is required as there is a change in delegate contract. 
</blockquote></details>

Done.
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java` line 68 at r7](https://reviewable.io/reviews/harness/harness-core/49204#gh-1233607258:-NYHXiIp9uj9DSc2uPMY:b-55up9s) ([raw file](https://github.com/harness/harness-core/blob/dd0ee2e587bc13a7177f8f15b5ba7d8c6dfa9859/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java#L68)):*
<details><summary><i>Previously, puthrayaharness (Deepak Puthraya) wrote…</i></summary><blockquote>

How will headers impact this check?
</blockquote></details>

The headers check will be used to handle if we are passing the headers for capability check and the ignoreResponseCode is to return valid capability if status code is less than 500, so we are having a combination of all the cases

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)* status: Some of the checks are too old. 1 change requests. 6. unresolved discussions. 

___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java` line 51 at r7](https://reviewable.io/reviews/harness/harness-core/49204#gh-1233605617:-NYH_6P712tCAJMz50dC:bm7pesk) ([raw file](https://github.com/harness/harness-core/blob/dd0ee2e587bc13a7177f8f15b5ba7d8c6dfa9859/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java#L51)):*
<details><summary><i>Previously, puthrayaharness (Deepak Puthraya) wrote…</i></summary><blockquote>

This whole code feels way too cluttered. There is too much nesting. Can we refactor the code for better readability? Consider having private methods.
</blockquote></details>

The nesting is due to FF added, we are planning to remove both of these FF so would refactor at that time

**Review by puthrayaharness — APPROVED** [review]

**Review by sahilhindwani-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49204)* status: Some of the checks are too old. 5. unresolved discussions.

**Pull request merged**